### PR TITLE
build typescript types

### PR DESCRIPTION
**Pull request by Marr11317**
The types files (`*.d.ts`) will are created when the project is built, but ignored by git. They are still included in the npm bundle (since all files in the `lib` folder are included)

### CODE REVIEW

**Comment by no-chris** [comment]
Thanks for this!

Could you explain a bit the use case? I'm a bit confused but that might be because of my total typescript ignorance.
I would have assumed that this PR was about adding TS types to the public API of the lib, so if you want to use it in a TS codebase you would benefit from all the TS addons. And this is what I seem to understand looking at the changes you made.

But in that case, why not commit the resulting file? It's unlikely that someone would include the package and then build it only to get the type? Or is it best practice to do it like that? Thanks

**Comment by Marr11317** [comment]
When you publish to npm, what you publish is what is indicated in your package.json as `files`. It has nothing to do with git. ie: npm doesn't read your gitignore to know what to include and what to ignore: it *only* reads your package .json.

I read somewhere that committing your lib folder wasn't a good idea. That's because you don't have to. As long as the necessary files are there when you call npm publish, they will be pushed to the registry.

**Comment by Marr11317** [comment]
I just converted it to a draft PR, because it looks like only the `parserFactory` and the `rendererfactory` are declared... I'll check if it can export your other types too (like `Chord`, `ChordInput`, `FormattedChord`, `NormalizedChord`)

**Comment by no-chris** [comment]
> I just converted it to a draft PR, because it looks like only the `parserFactory` and the `rendererfactory` are declared... I'll check if it can export your other types too (like `Chord`, `ChordInput`, `FormattedChord`, `NormalizedChord`)

That would be amazing! 👍

**Comment by no-chris** [comment]
> When you publish to npm, what you publish is what is indicated in your package.json as `files`. It has nothing to do with git. ie: npm doesn't read your gitignore to know what to include and what to ignore: it _only_ reads your package .json.
> 
> I read somewhere that committing your lib folder wasn't a good idea. That's because you don't have to. As long as the necessary files are there when you call npm publish, they will be pushed to the registry.

Right

So either the release script enforces the build before calling npm release
Either the lib folder is committed with everything in it

I think I don't mind committing the lib folder. I understand it bloats the repo and that it's not best practice but at the same time, it brings me better better confidence over what is published to npm. Otherwise, I don't really know what is happening there and in order to check what has been released, I need to download the repo and dig into the node_modules folder. So I'm fine to commit the `*.d.ts` file.

**Comment by Marr11317** [comment]
Here it is, with .d.ts files included. The problem is that typescript generates a lot of seemingly useless files, and I couldn't find a way to strip this to the minimum.

File: src/index.js [hunk 1]
```diff
@@ -2,3 +2,21 @@ import chordParserFactory from './parser/chordParserFactory';
...
+ * @export
+ */
+/**
+ * @typedef { import('./parser/chordParserFactory').FormattedChord } FormattedChord
+ * @export
+ */
+/**
+ * @typedef { import('./parser/chordParserFactory').NormalizedChord } NormalizedChord
+ * @export
+ */
```
Comment (no-chris) [inline src/index.js:22]: I'm puzzled. Why is this necessary if all the files in the repo get their corresponding d.ts file? Maybe just adding @export to the original declaration would do the trick?

Comment (Marr11317) [reply src/index.js:22]: Unfortunately, this is what I then get:
image

**Comment by Marr11317** [comment]
I don't like polluting the repo with so many files. The best option for now might be to simply close this PR, I'll use my own version of it when I need it, and if another user asks for the same, we'll reopen it. 
Does that make sense to you?

**Comment by no-chris** [comment]
That's indeed more files than I expected, I assumed only one would be created.

Let me test it locally and play around with it a bit first

File: src/index.js (hunk 1 above)
Comment (no-chris) [reply src/index.js:22]: But `Chord` should not be exported from the main entry point.

However, does your editor recognize `parsed` of being of type `Chord` ?

```javascript
const parseChord = chordParserFactory();
const parsed = parseChord('Cm7');
```

I tried on my own and it seems to work out of the box with the jsdoc annotations only

**Comment by no-chris** [comment]
According to https://www.typescriptlang.org/docs/handbook/declaration-files/dts-from-js.html
``` TypeScript supports most JSDoc tags```

So I tried a separate TS project, with `chord-symbol` on the `v1.0.0` branch.

Here is what I get:
Screenshot 2020-10-16 at 20 17 26

To me, it looks like typescript is doing its magic and the proper type inferences? The editor is `PHP Storm`, it might work differently in `VSCode`, though

**Comment by no-chris** [comment]
And the same code in a `.js` file instead of a `.ts` file, so there is definitely some typescript magic happening here since the type of `formatted` is not detected

Screenshot 2020-10-16 at 20 22 38

**Comment by no-chris** [comment]
Code completion works in `.ts`

Screenshot 2020-10-16 at 20 24 19

And not in `.js`

Screenshot 2020-10-16 at 20 24 29

So I wonder if anything is needed at all since my chordsymbol is on `v1.0.0` ?

**Comment by Marr11317** [comment]
Indeed, the type can be instantiated, but can't be referenced. It works with me too on vscode for completion.
image
The problem is about refering to the type: 
image
The type cannot be called directly.

**Comment by no-chris** [comment]
Got it. Same issue here on PHPStorm.

I'll keep trying for a while. In the meantime I'll release 1.0.0 and this can go in the next release.

**Comment by Marr11317** [comment]
Ok. There might be no need for d.ts files though if jsdoc comments may be read directly from js files.

**Comment by no-chris** [comment]
hu? I think this PR got automatically closed on deleting the 1.0.0 branch, sorry about that that was not intentional.

I like the idea of the feature and I'll keep on investigating it
